### PR TITLE
Use a patched go1.7.1 for building linux/arm

### PR DESCRIPTION
**Pull request by luxas**
Fixes: #29904

Right now, linux/arm is broken because of an internal limitation in Go.
I've filed an issue for it here: https://github.com/golang/go/issues/17028

The affected binaries of this limitation are hyperkube and kube-apiserver, which are the largest binaries.

And when we now have a patched go 1.7.1 version for building "unsupported" but important architectures (ref: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/multi-platform.md), we should also include the patch for ppc64le and start building ppc64le again. 

As soon as @laboger has the patch I need up on Github, I'll include ppc64le to this PR and we'll merge it

TODO:
- [ ] ~~Update the PR with patches for ppc64le at the same time @luxas~~
- [x] Push the new kube-cross image @ixdy 
- [x] Run a full `make release` before to verify nothing breaks @luxas + @ixdy 
- [ ] Cherrypick into the 1.4 branch @luxas + (who?)

@lavalamp @smarterclayton @ixdy @rsc @davecheney @wojtek-t @jfrazelle @bradfitz @david-mcmahon @pwittrock

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32517)

### CODE REVIEW

**Comment by luxas** [comment]
It'll fail until the image is pushed...

File: build/build-image/cross/VERSION
Comment (ixdy) [inline build/build-image/cross/VERSION]: this version tag is a bit misleading.

Comment (luxas) [inline build/build-image/cross/VERSION]: Hmm, what do you mean?
It's still 1.6.3 for official builds

Comment (ixdy) [inline build/build-image/cross/VERSION]: yeah, but it has a patched 1.7.1 Go in there too...

Comment (luxas) [inline build/build-image/cross/VERSION]: Yep, do you have a naming suggestion?

Comment (luxas) [inline build/build-image/cross/VERSION]: Ideally, I would have gone with a patched 1.6.3, but I'm not confident enough that it would work, since both CLs I've got are targeted at HEAD
I've already back-ported them once

Comment (ixdy) [inline build/build-image/cross/VERSION]: maybe we wait for #28742?

Comment (luxas) [inline build/build-image/cross/VERSION]: No, this should be cherrypicked into v1.4.
I think v1.6.3-7 is completely fine for now, it's just a new version of the image.

But we should make it a thing to document what has changed for each `vX.Y.Z-W` release, though

**Comment by foxish** [comment]
@luxas We will need to get this patch in by tomorrow evening at the latest if we want it cherrypicked into v1.4.

**Comment by eparis** [comment]
@pwittrock 
I'll be perfectly honest if APIServer has been broken on ARM and PCC64LE and we just found out now I do not think we should cherrypick to 1.4 and should instead evict the 'feature' from 1.4. It doesn't seem that it was complete or tested anywhere near on time.

Even if this merges I do not plan to cherry-pick. If others would like to disagree with my view please speak up.

**Comment by pwittrock** [comment]
@eparis sgtm

**Comment by luxas** [comment]
I've been testing the ppc64le stuff today, and it's not stable, so I'm gonna omit that.

ARM however is very important for me. This PR fixes that with no issues, and I'm begging that you're gonna cherrypick that. It has been working since v1.0.1, just to now hit a small linker issue, which I'm fixing here.

I have this PR ready for merge and cherrypick in a few hours.

**Comment by laboger** [comment]
@luxas can you give me more information on what kinds of stability issues you are seeing with ppc64le.  Are there some tests I can try.  Thanks.

**Comment by luxas** [comment]
@pwittrock @eparis This is ready for merge and cherrypick now

But first you have to run `make -C build/build-image/cross push`
cc @ixdy @david-mcmahon 

Please do this, now it's quite a small change, the only thing it does is fixing the linker issue which showed up quite recently on arm.

I'm totally ok with @eparis concern that ppc64le has been too little tested, and nuked it from this PR

I'm also fixing an issue that was caused by a wrong revert: https://github.com/kubernetes/kubernetes/pull/32404#discussion_r78461035

**Comment by ixdy** [comment]
if @pwittrock and/or @eparis LG, I can push the crossbuild image so we can test in this PR.

we should probably change the description/release note to note that `linux/arm` is being build with patched go1.7.1, not go1.6.3.

**Comment by pwittrock** [comment]
@luxas How about we plan this for 1.4.1?  There are a number of things going into the point release that are important.  Right now there are several open issues that we cannot launch without, so unless this should be one of them, we shouldn't cherrypick this until things are stable.

**Comment by eparis** [comment]
Isn't the amount of testing with 1.7.1 ~0? I do not want to put the kubernetes name behind something, even in 1.4.1 that seems like this. I feel like we missed 1.4 and should target 1.5. Lots of things are in this position. They are done and merged but weren't done in time for 1.4....

**Comment by luxas** [comment]
@eparis So instead you want to put your name behind this code?
```console
$ hyperkube
unexpected fault address 0x40d62ec
fatal error: fault
[signal 0xb code=0x2 addr=0x40d62ec pc=0x40d62ec]

goroutine 1 [running, locked to thread]:
runtime.throw(0x2cc9010, 0x5)
    /usr/local/go/src/runtime/panic.go:547 +0x78 fp=0x1482beec sp=0x1482bee0
runtime.sigpanic()
    /usr/local/go/src/runtime/sigpanic_unix.go:27 +0x280 fp=0x1482bf18 sp=0x1482beec
k8s.io/kubernetes/vendor/github.com/docker/engine-api/types/versions.init()
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/docker/engine-api/types/versions/compare.go:62 +0x4c fp=0x1482bf20 sp=0x1482bf1c
k8s.io/kubernetes/vendor/github.com/docker/engine-api/types/filters.init()
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/docker/engine-api/types/filters/parse.go:295 +0x5c fp=0x1482bf34 sp=0x1482bf20
k8s.io/kubernetes/vendor/github.com/docker/engine-api/types.init()
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/docker/engine-api/types/types.go:473 +0x5c fp=0x1482bf38 sp=0x1482bf34
k8s.io/kubernetes/pkg/credentialprovider.init()
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/credentialprovider/provider.go:123 +0x7c fp=0x1482bf6c sp=0x1482bf38
k8s.io/kubernetes/pkg/credentialprovider/aws.init()
and so on...
```

This is a **bug fix**, not a feature. 

> so unless this should be one of them,

@pwittrock This should definitely be one of them.

Please, can I get some understanding here? I've spent more than a year of **my spare time** on hacking Kubernetes (maintainer since April), and I'm the one that has introduced Kubernetes to multiple architectures. See for more information: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/multi-platform.md

Kubernetes has worked on ARM thanks to my efforts since v1.0.1, and has been built on releases since v1.2.

Now we're standing in front of a small linker issue in Golang. I've now fixed this issue by back-porting a CL to go1.7.1. **This is 1000x better than a binary that's segfaulting**. We're going to switch to go1.7.1 very soon for master as well, and there shouldn't be any significant changes.

@eparis If you feel very strongly, I can try to backport the CL to go1.6.3 as well, but IMO that would be **far more risky**.

Sorry for confusing you when I was talking about ppc64le here, that's not for the v1.4 branch.

I can personally take the blame for making Kubernetes work on arm, arm64 and partially ppc64le.

**Comment by foxish** [comment]
@luxas Let us get the tests passing and get it merged into master first. What is the next step for that? Once it is merged, we can evaluate whether it should be cherrypicked.

**Comment by luxas** [comment]
@ixdy should push the image and retrigger the tests

**Comment by lavalamp** [comment]
Wow, @luxas is definitely right that this is a bug fix and not a new feature. However building a patched version of go1.7 to build the arm distribution is not very appealing, either-- we haven't run with go1.7 and don't really know if there'd be any problems with it.

I can't really think of a better alternative, though. I guess @pwittrock will have to make the call.

**Comment by eparis** [comment]
how about a different question, what is changed between kube 1.3 and kube 1.4 which caused 1.6.3 (which apparently previously had a working API Server) such that we need to do this? Also of interest is why is it only being discovered now.

**Comment by pwittrock** [comment]
+What @eparis said

Also:
- Can this affect anything besides arm (e.g. could this affect the non-arm builds)?
- Does this only impact the arm tar.gz binaries that we release, or are there other arm related things it can impact?
- Do we need to support arm binaries at the time we release 1.4, or can we follow up with this later?
- What SLA do we provide for arm, and where is it published?
- How do we test and verify arm releases?
- If we don't accept this for 1.4.0 what are our options?  Simply not release arm binaries for 1.4.0?  Is there anything else that would break if we took that option?  Who is depending on these being published, and why is publishing them impactful enough to warrant slipping the release?

**Comment by luxas** [comment]
@eparis The following happened:
`v1.4.0-alpha.2` hyperkube broke. I opened #29904.
I thought the problem was the inits, so I opened #31431 (which is important anyway) and #31432 as a fix.
It didn't fix hyperkube, so we decided that it's not critical

When `v1.4.0-alpha.3` was released, `kube-apiserver` broke, which is the **really big** problem.
We can make it without hyperkube, but without the apiserver one can't do stuff.
So I opened https://github.com/golang/go/issues/17028 on the golang side this time, and it did recieve more attention.

Some of the guys there came up with a fix for Go tip, which I now am backporting to 1.7.1 (which already is risky, but probably less risky than using Go tip)

The problem is just that Kubernetes grew too big for the arm linker.

@lavalamp I definitely agree that this is not appealing nor ideal. 
But this is a way of getting at least kube-apiserver working for v1.4, and with some luck, hyperkube as well.

If you come up with some better solution, I'm all ears, but the timeframe is, as @eparis pointed out, tight.
This is a way of at least making the kube-apiserver

**Comment by luxas** [comment]
Seems like I and @pwittrock commented at the same time:

> Can this affect anything besides arm (e.g. could this affect the non-arm builds)?

No I don't think so

> Does this only impact the arm tar.gz binaries that we release, or are there other arm related things it can impact?

Yes

> Do we need to support arm binaries at the time we release 1.4, or can we follow up with this later?

Yes, we should aim to have non-segfaulting binaries on such a release

> What SLA do we provide for arm, and where is it published?

https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/multi-platform.md
Currently on a best-effort basis

> How do we test and verify arm releases?

Some guys have disabled `make release` on PR presubmit which turned out to be a disaster (delayed v1.4.0-alpha.3 with two weeks which led to this situation). 
We're trying to reenable it now.
Since it's hard for the community to run e2e tests, noone AFAIK is _continously running_ tests for master, but many people are using released versions of kubernetes on arm
But I'm manually testing arm from HEAD whenever I'm testing things.
Talked to @ixdy earlier if Google could setup a small nightly-build Jenkins arm CI, haven't really got a response yet

> If we don't accept this for 1.4.0 what are our options? Simply not release arm binaries for 1.4.0? Is there anything else that would break if we took that option? Who is depending on these being published, and why is publishing them impactful enough to warrant slipping the release?

We will, in either case, release arm binaries. The question is if they will segfault or not.

I'm not talking about slipping the release. Why would it if we're using this PR now?

**Comment by eparis** [comment]
lets say we go for it, step one is still make it work in master. As long as we are confident we can't break anything other than arm I'll likely try to pick it.

**Comment by luxas** [comment]
Ok, I'll update this finally tomorrow, it's very late here now...

**Comment by laboger** [comment]
_how about a different question, what is changed between kube 1.3 and kube 1.4 which caused 1.6.3 (which apparently previously had a working API Server) such that we need to do this? Also of interest is why is it only being discovered now._

This problem happens because of the program size on RISC platforms.  It is the same reason on ppc64le and arm.  What has changed is that more code was added between 1.3 and 1.4 to increase the size.  As long as the size is below the limit, it builds fine but once the limit is exceeded you can either get linker errors or runtime errors.  I don't know if you build for arm64 but I expect it should happen there too at some point.  This should get fixed in Go 1.8 but unfortunately the timing of this problem prevented a fix from being included in Go 1.7.  

I guess programs of this size were not expected or considered to be uncommon?

**Comment by ixdy** [comment]
> Ok, I'll update this finally tomorrow, it's very late here now...

@luxas Are you changing something in this? Or should I go ahead and push the image?

**Comment by pwittrock** [comment]
Thanks @luxas for your response.

Sounds like @eparis is on board with cherrypick this provided we get everything looking good in master.  Since he is doing the work for this I will back his decision.

> Do we need to support arm binaries at the time we release 1.4, or can we follow up with this later?

I am pushing on this because I want to fully understand what all of our options are.  Why would simply not publishing arm binaries fort 1.4.0 on Tuesday, and instead supporting it in 1.4.1 (possibly ~2 weeks) after the 1.4.0 launch be an unacceptable option?

> I'm not talking about slipping the release. Why would it if we're using this PR now?

You said in the corresponding issue that this is a launch blocker - e.g. we would block the launch on it - e.g. we would slip the release date for this instead of releasing without it.  I am not saying this alone will cause us to slip the release, but that is the severity you labelled it as and I wanted to confirm.  The discussion taking place on this thread has already diverted my attention from actually launch blocking issues and I need to prioritize it w.r.t. those issues going forward.  Every cherrypick we make introduces uncertainty, and the best tool are our disposal to minimize that uncertainty is to minimize the number of changes we are accepting into the branch.

**Comment by luxas** [comment]
@ixdy Yes, please push. Sorry for the confusion

**Comment by luxas** [comment]
ping @ixdy @eparis @david-mcmahon Can we proceed?
The local time for me is already 23:10, so I have to go to sleep soon.

This is ready for push, merge and cherrypick. It fixes the apiserver on arm.

File: hack/lib/golang.sh
Comment (ixdy) [inline hack/lib/golang.sh]: will we want to revert this at some point?

Comment (luxas) [inline hack/lib/golang.sh]: Yes, when things are even more stable, when I've had more time to dig into these cherrypicks, got more help from the Go team, and when we're using go1.7.1 at master.

I'm doing this to reduce the risk; @pwittrock and everyone wants as low risk as possible.

The priority is getting the apiserver working

File: cluster/images/hyperkube/static-pods/etcd.json
Comment (ixdy) [inline cluster/images/hyperkube/static-pods/etcd.json]: should we be worried about using 2.2.5 here but 2.2.1 everywhere else?

Comment (luxas) [inline cluster/images/hyperkube/static-pods/etcd.json]: See discussion here: https://github.com/kubernetes/kubernetes/pull/32404#r78450928

It's safe

**Comment by eparis** [comment]
i'll take a look tomorrow. no, I'm not going to merge today/tonight.

**Comment by eparis** [comment]
way to many build failures (so maybe ixdy can push his part today and we'll be better able to assess for 1.4 tomorrow)

**Comment by luxas** [comment]
> way to many build failures (so maybe ixdy can push his part today and we'll be better able to assess for 1.4 tomorrow)

all failures right now are because of the nonexistent image. 
@ixdy is pushing... and should retrigger the tests when done

**Comment by ixdy** [comment]
Tested `make release` with `gcr.io/google_containers/kube-cross:v1.6.3-7` locally, and pushed up to gcr.io.
The `linux/amd64` build is still using go1.6.3 as expected:
```shell
bin/linux/amd64 $ ./kubectl version
Client Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-alpha.0.800+41359888807f66-dirty", GitCommit:"41359888807f66f994334ddf66fd0b6582bceb26", GitTreeState:"dirty", BuildDate:"2016-09-14T22:13:51Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```
@k8s-bot test this please, issue #IGNORE

**Comment by foxish** [comment]
@k8s-bot test this please, issue #IGNORE

**Comment by luxas** [comment]
@k8s-bot test this please, issue #IGNORE

**Comment by ixdy** [comment]
@k8s-bot test this please, issue #IGNORE

**Comment by luxas** [comment]
@k8s-bot e2e test this please github issue: #IGNORE

**Comment by luxas** [comment]
@ixdy Please pull the new image on the kubemark nodes in order to make it pass, or switch to auto-pull if image does not exist.

Or someone just LGTM (kubemark isn't blocking)

**Comment by ixdy** [comment]
@k8s-bot kubemark test this please, issue #IGNORE

**Comment by ixdy** [comment]
I think the issue is that kubemark was triggered before I pushed the new image, and the 'test this' string apparently doesn't re-trigger it.

**Comment by luxas** [comment]
Okay, sounds logical. 
If we can get this into master today, it would be awesome, so @eparis can cherrypick tomorrow or maybe even today in your timezone.

File: hack/lib/golang.sh
Comment (ixdy) [inline hack/lib/golang.sh]: this might cause an issue if we get into this branch not inside the build image, though I think that'd only happen if you try to build k8s while on linux/arm.

**Review by ixdy — APPROVED** [review]
I'll let one more person LG and tag.

File: hack/lib/golang.sh
Comment (luxas) [inline hack/lib/golang.sh]: Yes, building **on** linux/arm with normally released Go is obviously broken (where the patched Golang doesn't exist)
Not much we can do about that for the meantime.

**Comment by luxas** [comment]
ping @david-mcmahon @foxish @eparis @pwittrock @lavalamp 
Someone please apply the lgtm label, @ixdy approved it
(Of course I can do it myself but seems like @ixdy want one more reviewer)

**Comment by luxas** [comment]
Also PTAL @vishh ASAP

**Comment by vishh** [comment]
This largely LGTM.

**Comment by luxas** [comment]
@k8s-bot unit test this please github issue: #IGNORE

**Comment by pwittrock** [comment]
Given the recent instability we have uncovered, and how late this is coming in, I am assuming this is something we will follow up on after 1.4.0 is out.

**Comment by lavalamp** [comment]
It's not merging on account of the unit test fail.

**Comment by lavalamp** [comment]
I think if you rebase that flake will go away.

**Comment by goltermann** [comment]
@k8s-bot unit test this please github issue: #IGNORE

**Comment by luxas** [comment]
@pwittrock Just two comments.

> Given the recent instability we have uncovered, 

It's totally unrelated unit flakes, not

> and how late this is coming in,

The PR was ready when you said it should be ready. As the author of the PR, I can only wait for you to push the image and LGTM it. That should be hold separately to when I had the PR ready for you

@eparis Another thing that you should consider; if you cherrypick this, _test targets will not be built for arm_, which _decreases_ the risk for things to break for the moment.

**Comment by jessfraz** [comment]
I believe we can remove all this now if we switch to go 1.7.3

**Comment by jessfraz** [comment]
I'll open a pr and have @luxas verify :)

**Comment by luxas** [comment]
Unfortunately I do not think go1.7.3 fixes the issue.
Changelog for v1.7.3: https://github.com/golang/go/issues?q=milestone%3AGo1.7.3
Issue with the linker: https://github.com/golang/go/issues/17028
Commits needed: https://github.com/golang/go/commit/c7e855658d73b85f345c9a0ac81de42acad7ae9b and https://github.com/golang/go/commit/7c431cb7f9780fcaf58b9ef07028d5129e1e5fe7

What do you think?

**Comment by laboger** [comment]
The fix for very large programs on ppc64le or arm was not allowed to be backported to go 1.7.3.

**Comment by jessfraz** [comment]
ah bummer nevermind then, we might as well update to 1.7.3 but I guess we
can't remove the patches yet

**Comment by luxas** [comment]
That's right, but when go1.8 is released, we may reenable ppc64le and remove the patches for arm :)

Do you know when go1.8 is supposed to be released?

**Comment by laboger** [comment]
Yes, fixes for both arm and ppc64le for this problem are upstream and will be in Go 1.8, scheduled to be released in Feb. 2017. 

There is a patch for go 1.7.x that will work on ppc64le if you enable external linking for all binaries.

**Comment by luxas** [comment]
When will the RCs/betas start to show up?

Because I'd definitely like to get arm external linking working before v1.5...
Sounds like I have to cherry-pick those changes upon go1.7.3 then for v1.5

@laboger If you send me a complete patch that is applicable upon go1.7.3 that fixes all of the current ppc64le problems, I may include it as well so we can start building ppc64le as well.

**Comment by laboger** [comment]
@luxas
I can see that go1.7beta1 was available in early June, and go1.7 released in August.

The only ppc64le problem I am aware of is the one due to the size of the program.  What other problems have you seen?

**Comment by jessfraz** [comment]
In about a month IIRC

**Comment by laboger** [comment]
@luxas
We have repository with golang releases + the fixes we have pushed upstream.
The branch ibm-1.7 from https://github.com/powertechpreview/go.git has the fix for big text sections when using external linking.  You can cherry pick these commits on top of the go1.7.3 branch:
git cherry-pick d6beea7f
git cherry-pick a5b97a84

I just did this and it all worked for me and I was able to successfully build kubernetes latest.  These are the same instructions I gave you before so if there is a problem with part of it let me know.  
KUBE_BUILD_PPC64LE=y KUBE_BUILD_GOLDFLAGS='-linkmode=external' make 

Using -linkmode=external is only a temporary workaround until Go 1.8 is available.

**Pull request merged**
